README.md
# innosoft_simpletooltip

Plugin aims to help produce an easily usable implementation of tooltip

## Usage
To use this plugin, add ```innosoft_simpletooltip``` as a [dependency in your pubspec.yaml](https://flutter.io/platform-plugins/).

### Example
<pre>import 'package:flutter/material.dart';<br />import 'package:innosoft_simpletooltip/innosoft_simpletooltip.dart';<br /><br />void main() {<br />  runApp(App());<br />}<br /><br />class App extends StatelessWidget {<br /><br />  @override<br />  Widget build(BuildContext context) {<br />    return MaterialApp(<br />      home: TooltipExampleStateful(),<br />    );<br />  }<br />}<br /><br />class TooltipExampleStateful extends StatefulWidget {<br />  TooltipExampleStateful({Key key}) : super(key: key);<br /><br />  @override<br />  _TooltipExampleStateful createState() =&gt; _TooltipExampleStateful();<br />}<br /><br />class _TooltipExampleStateful extends State&lt;TooltipExampleStateful&gt; {<br /><br />  double marginTop = 20;<br />  SoftSimpleTooltip tooltipUp, tooltipDown;<br />  bool clsTooltipUp = false;<br />  bool clsTooltipDown = false;<br /><br />  @override<br />  void initState() {<br />    super.initState();<br />    tooltipUp = SoftSimpleTooltip(<br />        content: Container(<br />          //height: 100.0,<br />            child:Text("Lorem ipsum dolor sit amet, consetetur sadipscingelitr, "<br />                "sed diam nonumy eirmod tempor invidunt ut laboreet dolore magna aliquyam erat, "<br />                "sed diam voluptua. At vero eos et accusam et justoduo dolores et ea rebum. ")<br />        ),<br />        popupDirection: TooltipDirection.up,<br />        backgroundColor: Colors.black,<br />        radius: 20,<br />        arrowHeight: 40,<br />        arrowWidth: 50,<br />        arrowPos: 200,<br />        tooltipStyle: TooltipStyle.border);<br />    tooltipDown = SoftSimpleTooltip(<br />        content: Container(<br />          //height: 100.0,<br />            child:Text("Lorem ipsum dolor sit amet, consetetur sadipscingelitr, "<br />                "sed diam nonumy eirmod tempor invidunt ut laboreet dolore magna aliquyam erat, "<br />                "sed diam voluptua. At vero eos et accusam et justoduo dolores et ea rebum. ")<br />        ),<br />        popupDirection: TooltipDirection.down,<br />        backgroundColor: Colors.black,<br />        radius: 20,<br />        arrowHeight: 40,<br />        arrowWidth: 50,<br />        arrowPos: 200,<br />        tooltipStyle: TooltipStyle.border);<br />  }<br /><br />  @override<br />  void dispose() {<br />    super.dispose();<br />  }<br /><br />  void showHideTooltipUp(){<br />    clsTooltipUp ? tooltipUp.showTooltip() : tooltipUp.closeTooltip();<br />    setState(() {<br />      clsTooltipUp = !clsTooltipUp;<br />    });<br />  }<br /><br />  void showHideTooltipDown(){<br />    clsTooltipDown ? tooltipDown.showTooltip() : tooltipDown.closeTooltip();<br />    setState(() {<br />      clsTooltipDown = !clsTooltipDown;<br />    });<br />  }<br /><br />  @override<br />  Widget build(BuildContext context) {<br />    return Scaffold(<br />        backgroundColor: Colors.white,<br />        appBar: AppBar(<br />          backgroundColor: Colors.white,<br />          elevation: 0.0,<br />          centerTitle: true,<br />          title: Text(<br />            'TOOLTIP',<br />            style: TextStyle(color: Colors.black),<br />          ),<br />        ),<br />        body:<br />        Column(<br />          //mainAxisAlignment: MainAxisAlignment.spaceAround,<br />            mainAxisSize: MainAxisSize.min,<br />            children: &lt;Widget&gt;[<br />              // Row 1<br />              Row(<br />                  crossAxisAlignment: CrossAxisAlignment.start,<br />                  children: &lt;Widget&gt;[<br />                    Expanded(<br />                        child: new Container(<br />                            margin: new EdgeInsets.fromLTRB(20, 20, 20, 0),<br />                            child: Row(<br />                              children: &lt;Widget&gt;[<br />                                Expanded(<br />                                  child: tooltipUp,<br />                                ),<br />                              ],<br />                            ))<br />                    ),<br />                  ]<br />              ),<br />              Row(<br />                  children: &lt;Widget&gt;[<br />                    Expanded(<br />                        child: new Container(<br />                            margin: new EdgeInsets.fromLTRB(20, 0, 20, 0),<br />                            child: RaisedButton(<br />                              onPressed: () {<br />                                //Navigator.pop(globalContext);<br />                                showHideTooltipUp();<br />                              },<br />                              child: Text('Show Hide Tooltip Up'),<br />                            ))<br />                    ),<br />                  ]<br />              ),<br />              Row(<br />                  children: &lt;Widget&gt;[<br />                    Expanded(<br />                        child: new Container(<br />                            margin: new EdgeInsets.fromLTRB(20, 20, 20, 0),<br />                            child: RaisedButton(<br />                              onPressed: () {<br />                                //Navigator.pop(globalContext);<br />                                showHideTooltipDown();<br />                              },<br />                              child: Text('Show Hide Tooltip Down'),<br />                            ))<br />                    ),<br />                  ]<br />              ),<br />              Row(<br />                  crossAxisAlignment: CrossAxisAlignment.start,<br />                  children: &lt;Widget&gt;[<br />                    Expanded(<br />                        child: new Container(<br />                            margin: new EdgeInsets.fromLTRB(20, 0, 20, 0),<br />                            child: Row(<br />                              children: &lt;Widget&gt;[<br />                                Expanded(<br />                                  child: tooltipDown,<br />                                ),<br />                              ],<br />                            ))<br />                    ),<br />                  ]<br />              ),<br />            ]<br />        )<br />    );<br />  }<br />}</pre>



CHANGELOG.md
## [0.0.1] - 06/03/2020.

* Inital publish package


test/innosoft_simpletooltip_test.dart
import 'package:flutter/material.dart';
import 'package:innosoft_simpletooltip/innosoft_simpletooltip.dart';

void main() => runApp(App());

class App extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: TooltipExampleStateful(),
    );
  }
}

class TooltipExampleStateful extends StatefulWidget {
  TooltipExampleStateful({Key key}) : super(key: key);

  @override
  _TooltipExampleStateful createState() => _TooltipExampleStateful();
}

class _TooltipExampleStateful extends State<TooltipExampleStateful> {

  double marginTop = 20;
  SoftSimpleTooltip tooltipUp, tooltipDown;
  bool clsTooltipUp = false;
  bool clsTooltipDown = false;

  @override
  void initState() {
    super.initState();
    tooltipUp = SoftSimpleTooltip(
        content: Container(
          //height: 100.0,
            child:Text("Lorem ipsum dolor sit amet, consetetur sadipscingelitr, "
                "sed diam nonumy eirmod tempor invidunt ut laboreet dolore magna aliquyam erat, "
                "sed diam voluptua. At vero eos et accusam et justoduo dolores et ea rebum. ")
        ),
        popupDirection: TooltipDirection.up,
        backgroundColor: Colors.black,
        radius: 20,
        arrowHeight: 40,
        arrowWidth: 50,
        arrowPos: 200,
        tooltipStyle: TooltipStyle.border);
    tooltipDown = SoftSimpleTooltip(
        content: Container(
          //height: 100.0,
            child:Text("Lorem ipsum dolor sit amet, consetetur sadipscingelitr, "
                "sed diam nonumy eirmod tempor invidunt ut laboreet dolore magna aliquyam erat, "
                "sed diam voluptua. At vero eos et accusam et justoduo dolores et ea rebum. ")
        ),
        popupDirection: TooltipDirection.down,
        backgroundColor: Colors.black,
        radius: 20,
        arrowHeight: 40,
        arrowWidth: 50,
        arrowPos: 200,
        tooltipStyle: TooltipStyle.border);
  }

  @override
  void dispose() {
    super.dispose();
  }

  void showHideTooltipUp(){
    clsTooltipUp ? tooltipUp.showTooltip() : tooltipUp.closeTooltip();
    setState(() {
      clsTooltipUp = !clsTooltipUp;
    });
  }

  void showHideTooltipDown(){
    clsTooltipDown ? tooltipDown.showTooltip() : tooltipDown.closeTooltip();
    setState(() {
      clsTooltipDown = !clsTooltipDown;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        backgroundColor: Colors.white,
        appBar: AppBar(
          backgroundColor: Colors.white,
          elevation: 0.0,
          centerTitle: true,
          title: Text(
            'TOOLTIP',
            style: TextStyle(color: Colors.black),
          ),
        ),
        body:
        Column(
          //mainAxisAlignment: MainAxisAlignment.spaceAround,
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              // Row 1
              Row(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    Expanded(
                        child: new Container(
                            margin: new EdgeInsets.fromLTRB(20, 20, 20, 0),
                            child: Row(
                              children: <Widget>[
                                Expanded(
                                  child: tooltipUp,
                                ),
                              ],
                            ))
                    ),
                  ]
              ),
              Row(
                  children: <Widget>[
                    Expanded(
                        child: new Container(
                            margin: new EdgeInsets.fromLTRB(20, 0, 20, 0),
                            child: RaisedButton(
                              onPressed: () {
                                //Navigator.pop(globalContext);
                                showHideTooltipUp();
                              },
                              child: Text('Show Hide Tooltip Up'),
                            ))
                    ),
                  ]
              ),
              Row(
                  children: <Widget>[
                    Expanded(
                        child: new Container(
                            margin: new EdgeInsets.fromLTRB(20, 20, 20, 0),
                            child: RaisedButton(
                              onPressed: () {
                                //Navigator.pop(globalContext);
                                showHideTooltipDown();
                              },
                              child: Text('Show Hide Tooltip Down'),
                            ))
                    ),
                  ]
              ),
              Row(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    Expanded(
                        child: new Container(
                            margin: new EdgeInsets.fromLTRB(20, 0, 20, 0),
                            child: Row(
                              children: <Widget>[
                                Expanded(
                                  child: tooltipDown,
                                ),
                              ],
                            ))
                    ),
                  ]
              ),
            ]
        )
    );
  }
}


example/android/app/src/main/kotlin/app/test/com/example/MainActivity.kt
package app.test.com.example

import io.flutter.embedding.android.FlutterActivity

class MainActivity: FlutterActivity() {
}


lib/innosoft_simpletooltip.dart
library innosoft_simpletooltip;

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

enum TooltipDirection { up, down}

enum TooltipStyle { fill, border}

var globalContext;

double oriLayoutHeight, layoutHeight;

bool _isVisible, ftVisible;

class SoftSimpleTooltip extends StatefulWidget {
  Widget content;

  TooltipDirection popupDirection;

  TooltipStyle tooltipStyle;

  Color backgroundColor;

  bool hasShadow;

  double radius,
  //arrowDirection,
      arrowHeight,
      arrowWidth;
  //tt_gravity,;
  double arrowPos,
      heightWall;

  _SoftSimpleTooltip stp = new _SoftSimpleTooltip();

  SoftSimpleTooltip(
      {
        Key key,
        @required this.content, // The contents of the tooltip.
        @required this.popupDirection,
        @required this.backgroundColor,
        @required this.radius,
        @required this.arrowHeight,
        @required this.arrowWidth,
        @required this.arrowPos,
        //@required this.tt_gravity,
        @required this.tooltipStyle,
        this.hasShadow
      }) : super(key: key);

  @override
  _SoftSimpleTooltip createState() => stp;

  void closeTooltip(){
    stp.closeTooltip();
  }

  void showTooltip(){
    stp.showTooltip();
  }
}

class _SoftSimpleTooltip extends State<SoftSimpleTooltip> {
  final String tag = "SoftSimpleTooltip";

  @override
  void initState() {
    super.initState();
    _isVisible = true;
    ftVisible = true;
  }

  @override
  void didChangeDependencies() {
  }

  @override
  void didUpdateWidget(Widget oldWidget)
  {
    super.didUpdateWidget(oldWidget);
  }

  @override
  void deactivate() {
  }

  @override
  void dispose()
  {
    super.dispose();
  }

  void printWrapped(String tag, String text) {
    final pattern = RegExp('.{1,800}'); // 800 is the size of each chunk
    pattern.allMatches(tag+' -> '+text).forEach((match) => print(match.group(0)));
  }

  void showTooltip(){
    setState(() {
      _isVisible = true;
      layoutHeight = oriLayoutHeight;
    });
  }

  void closeTooltip(){
    setState(() {
      layoutHeight = 0;
    });
    new Future.delayed(new Duration(milliseconds: 180), () {
      setState(() {
        _isVisible = false;
      });
    });
  }

  _SoftSimpleTooltip();

  @override
  Widget build(BuildContext context) {
    globalContext = context;
    // TODO: implement build
    return new Container(
      child:
      new Container(
        margin: new EdgeInsets.fromLTRB(20, 20, 20, 20),
        child: Row(
          children: <Widget>[
            Expanded(
              child: Visibility (
                visible: _isVisible,
                child:AnimatedContainer(
                  height: layoutHeight,
                  child: CustomPaint(
                    child: Container(
                        margin: _getContainerMargin(),
                        child:widget.content
                    ),
                    painter: PathPainter(
                        widget.popupDirection,
                        widget.backgroundColor,
                        widget.radius,
                        widget.arrowHeight,
                        widget.arrowWidth,
                        widget.arrowPos,
                        widget.tooltipStyle),
                  ),
                  duration: Duration(milliseconds: 200),
                  // Provide an optional curve to make the animation feel smoother.
                  curve: Curves.easeInOut,
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }

  EdgeInsets _getContainerMargin() {
    switch (widget.popupDirection) {
    //
      case TooltipDirection.down:
        double top = widget.arrowHeight+widget.radius;
        return EdgeInsets.fromLTRB(widget.radius, top, widget.radius, widget.radius);

      case TooltipDirection.up:
        double bottom = widget.arrowHeight+widget.radius;
        return EdgeInsets.fromLTRB(widget.radius, widget.radius, widget.radius, bottom);

      default:
        throw AssertionError(widget.popupDirection);
    }
  }
}

class PathPainter extends CustomPainter {

  double radius, arrowHeight, arrowWidth;
  double arrowPos;
  Color backgroundColor;
  TooltipDirection popupDirection;
  TooltipStyle tooltipStyle;

  PathPainter(
      TooltipDirection popupDirection,
      Color backgroundColor,
      double radius,
      double arrowHeight,
      double arrowWidth,
      double arrowPos,
      TooltipStyle tooltipStyle)
  {
    this.popupDirection = popupDirection;
    this.backgroundColor = backgroundColor;
    this.radius = radius;
    this.arrowHeight = arrowHeight;
    this.arrowWidth = arrowWidth;
    this.arrowPos = arrowPos;
    this.tooltipStyle = tooltipStyle;
  }

  @override
  void paint(Canvas canvas, Size size) {

    double w = size.width;
    double h = size.height;
    if(ftVisible) {
      oriLayoutHeight = h;
      ftVisible = false;
    }
    layoutHeight = h;

    Paint paint = Paint();
    paint.color = backgroundColor;

    Path path = Path();

    if(tooltipStyle == TooltipStyle.fill) {
      paint.style = PaintingStyle.fill;
    }
    else{
      paint.style = PaintingStyle.stroke;
      paint.strokeWidth = 4.0;
    }

    if(arrowHeight > (h/2))arrowHeight = h/2;
    if(arrowWidth > (w/2))arrowWidth = w/2;
    double ch = h/8;
    if(radius > ch)radius = ch;
    double p = arrowHeight;
    double p2 = arrowWidth/2;
    if(arrowPos > 0){
      if(arrowPos < radius)arrowPos = radius;
      else if(arrowPos > (w-(radius*4)))arrowPos = w-(radius*4);
    }
    double c = arrowPos <= 0 ? (w/2)-(p2) : arrowPos;
    c = c.abs();
    if(c < radius)c=radius;
    else {
      double limitRight = w - (radius * 2);
      if (c > limitRight) c = limitRight;
    }

    double c2 = c+p2;
    double c3 = c+(p2*2);
    double y = h-(h-p);

    if(popupDirection == TooltipDirection.down) {
      path.moveTo(0, y + radius);
      path.quadraticBezierTo(0, y, radius, y);
      path.lineTo(c, y);
      path.lineTo(c2, 0);
      path.lineTo(c3, y);

      path.lineTo(w - radius, y);
      path.quadraticBezierTo(w, y, w, y + radius);

      path.lineTo(w, h - radius);
      path.quadraticBezierTo(w, h, w - radius, h);

      path.lineTo(radius, h);
      path.quadraticBezierTo(0, h, 0, h - radius);

      path.lineTo(0, y + radius);
      canvas.drawPath(path, paint);
    }
    else {
      double posFromBottom = h-arrowHeight;
      path.moveTo(0, radius);
      path.quadraticBezierTo(0, 0, radius, 0);

      path.lineTo(w - radius, 0);
      path.quadraticBezierTo(w, 0, w, radius);

      path.lineTo(w, posFromBottom - radius);
      path.quadraticBezierTo(w, posFromBottom, w - radius, posFromBottom);

      path.lineTo(c3, posFromBottom);
      path.lineTo(c2, h);
      path.lineTo(c, posFromBottom);

      path.lineTo(radius, posFromBottom);
      path.quadraticBezierTo(0, posFromBottom, 0, posFromBottom-radius);

      path.lineTo(0, radius);
      canvas.drawPath(path, paint);
    }

  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:innosoft_simpletooltip/innosoft_simpletooltip.dart';

void main() {
  runApp(App());
}

class App extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: TooltipExampleStateful(),
    );
  }
}

class TooltipExampleStateful extends StatefulWidget {
  TooltipExampleStateful({Key key}) : super(key: key);

  @override
  _TooltipExampleStateful createState() => _TooltipExampleStateful();
}

class _TooltipExampleStateful extends State<TooltipExampleStateful> {

  double marginTop = 20;
  SoftSimpleTooltip tooltipUp, tooltipDown;
  bool clsTooltipUp = false;
  bool clsTooltipDown = false;

  @override
  void initState() {
    super.initState();
    tooltipUp = SoftSimpleTooltip(
        content: Container(
          //height: 100.0,
            child:Text("Lorem ipsum dolor sit amet, consetetur sadipscingelitr, "
                "sed diam nonumy eirmod tempor invidunt ut laboreet dolore magna aliquyam erat, "
                "sed diam voluptua. At vero eos et accusam et justoduo dolores et ea rebum. ")
        ),
        popupDirection: TooltipDirection.up,
        backgroundColor: Colors.black,
        radius: 20,
        arrowHeight: 40,
        arrowWidth: 50,
        arrowPos: 200,
        tooltipStyle: TooltipStyle.border);
    tooltipDown = SoftSimpleTooltip(
        content: Container(
          //height: 100.0,
            child:Text("Lorem ipsum dolor sit amet, consetetur sadipscingelitr, "
                "sed diam nonumy eirmod tempor invidunt ut laboreet dolore magna aliquyam erat, "
                "sed diam voluptua. At vero eos et accusam et justoduo dolores et ea rebum. ")
        ),
        popupDirection: TooltipDirection.down,
        backgroundColor: Colors.black,
        radius: 20,
        arrowHeight: 40,
        arrowWidth: 50,
        arrowPos: 200,
        tooltipStyle: TooltipStyle.border);
  }

  @override
  void dispose() {
    super.dispose();
  }

  void showHideTooltipUp(){
    clsTooltipUp ? tooltipUp.showTooltip() : tooltipUp.closeTooltip();
    setState(() {
      clsTooltipUp = !clsTooltipUp;
    });
  }

  void showHideTooltipDown(){
    clsTooltipDown ? tooltipDown.showTooltip() : tooltipDown.closeTooltip();
    setState(() {
      clsTooltipDown = !clsTooltipDown;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        backgroundColor: Colors.white,
        appBar: AppBar(
          backgroundColor: Colors.white,
          elevation: 0.0,
          centerTitle: true,
          title: Text(
            'TOOLTIP',
            style: TextStyle(color: Colors.black),
          ),
        ),
        body:
        Column(
          //mainAxisAlignment: MainAxisAlignment.spaceAround,
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              // Row 1
              Row(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    Expanded(
                        child: new Container(
                            margin: new EdgeInsets.fromLTRB(20, 20, 20, 0),
                            child: Row(
                              children: <Widget>[
                                Expanded(
                                  child: tooltipUp,
                                ),
                              ],
                            ))
                    ),
                  ]
              ),
              Row(
                  children: <Widget>[
                    Expanded(
                        child: new Container(
                            margin: new EdgeInsets.fromLTRB(20, 0, 20, 0),
                            child: RaisedButton(
                              onPressed: () {
                                //Navigator.pop(globalContext);
                                showHideTooltipUp();
                              },
                              child: Text('Show Hide Tooltip Up'),
                            ))
                    ),
                  ]
              ),
              Row(
                  children: <Widget>[
                    Expanded(
                        child: new Container(
                            margin: new EdgeInsets.fromLTRB(20, 20, 20, 0),
                            child: RaisedButton(
                              onPressed: () {
                                //Navigator.pop(globalContext);
                                showHideTooltipDown();
                              },
                              child: Text('Show Hide Tooltip Down'),
                            ))
                    ),
                  ]
              ),
              Row(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    Expanded(
                        child: new Container(
                            margin: new EdgeInsets.fromLTRB(20, 0, 20, 0),
                            child: Row(
                              children: <Widget>[
                                Expanded(
                                  child: tooltipDown,
                                ),
                              ],
                            ))
                    ),
                  ]
              ),
            ]
        )
    );
  }
}